Added DingTalk Support

chengyuhui opened the pull request:
## Description:
**Related issue (if applicable):** #200 

<!-- Have anything else to describe? Define it here -->
Added [DingTalk](https://dingtalk.com) Support with signature generation based on the original PR.

There seems to be ZERO documents on how to (or not to) correctly add new services, and a quick scan through the PR list showed that no one has ever tried to add new services except the owner, so I just did what I can do.

`dingtalk:// template/arg/func reference missing error.` is not a helpful message to me.

## New Service Completion Status
<!-- This section is only applicable if you're adding a new service -->
* [x] apprise/plugins/NotifyDingTalk.py
* [x] setup.py
    - add new service into the `keywords` section of the `setup()` declaration
* [x] README.md
    - add entry for new service to table (as a quick reference)
* [x] packaging/redhat/python-apprise.spec
    - add new service into the `%global common_description`

## Checklist
<!-- The following must be completed or your PR can't be merged -->
* [x] The code change is tested and works locally.
* [x] There is no commented out code in this PR.
* [x] No lint errors (use `flake8`)
* [ ] 100% test coverage (`FAILED test/test_api.py::test_apprise_details_plugin_verification - AssertionError: assert 'secret' in {'cto', 'format', 'fullpath', 'host', 'overflow', 'password', ...}`)

caronc commented:
First off; let me tell you this is an amazing contribution! Thank you so much :+1: :heart: 

You're getting stuck on a simple check i made for myself because i too make this mistake a lot.   The error isn't really telling. Basically the only step is that for everything you define in `NotifyDingTalk.init(functions)` you need to also identify in `NotifyDingTalk.template_tokens`.

Just at a glance, it looks like you defined `token`, `secret`, and `targets`.  However you do not handle `secrets`.

It's very important that you handle `NotifyDingTalk.url()` to rebuild this same URL. You also want to be able to parse `targets`, `secret`, and `token` in the `parse_url()` function.  At this time it looks like you only handle `targets` and `token`.

Perhaps `secret` isn't required at all? If that's the case you can just drop it from the `template_tokens` and you'll get a lot farther.

Let me know if you have any questions! 

Chris

chengyuhui commented:
`secret` is an optional parameter that I decided to put in the `user` part of the URL, it may or may not be needed depending on what is set up in their DingTalk UI.

So you mean that I can simply read from `self.user` without defining it?

caronc commented:
It might make more sense as a `?kwarg=` if it's not necessary or required to be specified.  I do acknowledge your point also that there is very lacking documentation on how to build a plugin. Kudos for all you've done though.  I'm actually going to be out of town for the next 2 days.  I'll be back at my desk (at home) on Monday.  I'll give you a hand to finish this off; you're so close. :slightly_smiling_face:

caronc commented:
Now that I've finally had a chance to sit down and look at your code; have you considered this [DingTalk PR](https://github.com/caronc/apprise/pull/200) I created back in June 2020? Just realizing now that you also made a comment on it too. That one was parked simply because I had no one to test it for me.

Is this just a copy/paste of that PR and you're trying to introduce the `secret` argument?  Either way, it seems like your close all the same. So we can use either/or branch. :slightly_smiling_face: 

As per this comment:
> There seems to be ZERO documents on how to (or not to) correctly add new services, and a quick scan through the PR list showed that no one has ever tried to add new services except the owner, so I just did what I can do.

I couldn't agree with you more! Kudos for making such a valiant effort of picking up the code and working with it though!  I will try to rough up some documentation.  But in short:
- You just need to create you just need to create a single notification in the `/plugins/NotifyServiceName.py`
- Make sure you call the class inside `NotifyServiceName` and inherit from `NotifyBase`
- From there you just need to at a bare minimum define:
  - **the class objects**:
    - `service_name`: A string that acts as a default descriptive name associated with the Notification
    - `service_url`: A string that identifies the platform/services URL.  This is used purely as meta data for those who seek it. But this field is required.
    - `secure_protocol`: A string (or can be a list of strings) identifying the scheme:// keyword.   So setting this to say `'bad'` means that this class you're creating will be instantiated if someone tries to access `bad://`
    - `setup_url` =  A string that identifies the URL a user can use to get information on how to use this Apprise Notification.  At this time I'm just creating URLs that point back to my GitHub Wiki page.
   - `templates`, `template_tokens`, and `template_args`.  It's best to look at all of the many services that exist to see how to prepare this.  But this is just a way of identifying what the class accepts as arguments. You can optionally identify some regular expressions to validate the arguments as well.  This is provided as both meta data to those requesting it (it plays a huge role in the API i built as the JSON output is based on this).  This data is also heavily validated using unit tests.  So if you identify something there that isn't defined in the `__init__()` of your class, or vs versa (you declare something in your `__init__()` and do not identify it here.  Then you'll get an error (tests will not pass).  The problem is the error is rather cryptic; (this is what is happening to you; you've identified `secret` in only one location)
  
  - **the functions**:
     - `__init__(self, *args, **kwargs)`: Define what is required to initialize your object/notification. Just make sure to cross reference it in the `template*` stuff (explained above).
     -`send(self, body, title='', *args, **kwargs)` at a bare minimum. See other Notify scripts as to how you can expand on this.  But just take the `body` and `title` and construct your message and send it.  
     - `url()`.  This function must be able to construct a URL that would re-generate a copy of the exact same object if passed into `parse_url()`
     - `parse_url(url)`: this is a **staticmethod** that parses the Apprise URL and breaks it into a dictionary of the components.  The dictionary it creates must map up to what the `__init__()` takes as it's arguments
     
  - **Putting it together**:
       ```python
       from Apprise.plugins import NotifyMyService
       import Apprise
       
       # Apprise is nothing but a manager of individual plugins
       a = Apprise()
       a.add('myscheme://details/?more=details&are=optional')
       
       # There would be one new service added to our manager now:
       assert(len(a), 1)

       # you can directly access the notification services if you wanted to this way:
       # index element 0 exists because we added it successfully above (assuming you properly
       # followed all the rules above):
       assert isinstance(a[0], NotifyMyService)
       
       # So we know we can access the notification, then this would create a second notification service:
       # The only thing add does is match the schema up with the class it should use and then call it's
       # NotifyServiceName.parse_url()
       
       # So parse_url() is in charge of preparing all of the arguments we can use to instantiate our object
       #  With that, it can then do Object(**parse_url_response)
       a.add(a[0].url())

       # Hopefully this is making sense so far.... But now we've called add() twice... so we'll ahve 2 entries
       # and if we built our 3 core functions (__init__, `url()` and `parse_url()` correctly, they should be almost
       # copies of one another (yet 2 instances)
       assert(len(a) == 2)
      
       # URLs are the same
       assert(a[0].url() == a[1].url())
      
       # that's really all there is too it... when you call `a.notify()`; there is some functions and tools
       # that handle some common things, but at the end of the day, it will call your `send()` function
       # you defined.
       ```
      Any other functions you want to define can be done to you hearts content (if it helps with organization, structure, etc)
      Just avoid conflicting with any function written in `NotifyBase()` and `URLBase()`

      If your service is really complex (and requires a lot of code), maybe it's easier to split your code into multiple files. This is how i handled the [NotifyFCM plugin i wrote](https://github.com/caronc/apprise/tree/master/apprise/plugins/NotifyFCM) which was based on Google's version.

You can have a look at the NotifyBase object and see all of the other entries you can define that Apprise can look after for you (such as restricting the message length, title length, handling TEXT -> Markdown, etc). You can also look at how other classes were built.

chengyuhui commented:
In case you haven't noticed it (it's probably been a long time), I took your code from that PR, rebased it to master, and added the secret handling feature into that code.

shiba2046 commented:
```
> apprise -vvv  dingtalk://dbade.../
2021-07-26 15:25:37,286 - ERROR - Unparseable DingTalk URL dingtalk://dbade..../
2021-07-26 15:25:37,286 - ERROR - You must specify at least one server URL or populated configuration file.
```
I tried to use the Dingtalk url scheme but was met with this problem. It seems it has problem to work from cli?

caronc commented:
Hi, I updated the original DingTalk branch (#200) i created over a year and a half a go with all of your adjustments (full credit to all source code there reguarding the signing/timestamp of the `{secret}` references there).   I also added all of the unit tests to cover everything (back to 100% test coverage).

Feel free to review/test it and/or copy and paste it all back into your branch. It honestly doesn't bother me to hand you the credit.  The main thing is getting rid of the pull request since it's been lingering for so long (untested). It's great to see you come along and help me get it into Apprise. :+1:

chengyuhui commented:
Okay, I'll have it tested when I'm back in my office tomottow.

caronc commented:
Closing as per MR #200
